File: Experiments/Run-time-Polymorphism_Inheritable_No-type-checking/include/duck.r
#ifndef DUCK_R
#define DUCK_R

#include <stdarg.h>

#define MAX_CHARS_NAME 10
#define MAX_CHARS_NAME_WITH_NUL (MAX_CHARS_NAME+1)

typedef struct Duck_Interface_Struct const * Duck_Interface;

typedef struct Duck_t
{
    Duck_Interface vtable;
    char name[MAX_CHARS_NAME_WITH_NUL];
} Duck_t, *Duck;

typedef struct Duck_Interface_Struct
{
    void (*show)( void * thisDuck );
    void (*deinit)( void * thisDuck );
    void (*destroy)( void * thisDuck );
} Duck_Interface_Struct;

void duckInit( Duck thisDuck, char * name );
void duckDeinit( void * thisDuck );

#endif // DUCK_R

File: 3_Run-time-Polymorphism/c_Making-mallard-objects/source/duck.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "assert.h"
#include "duck.h"
#include "duck_private.h"

p_Duck_t
duckNew( char * name )
{
    p_Duck_t newDuck = (p_Duck_t)calloc(1, sizeof(Duck_t));
    // TODO: Check for null pointer on malloc failure

    duckInit(newDuck, name);

    return newDuck;
}

void
duckQuack( p_Duck_t thisDuck )
{
    ASSERT(thisDuck && thisDuck->name);
    
    printf("\t%s: Quack!!\n", thisDuck->name);
}

void
duckShow( p_Duck_t thisDuck )
{
    ASSERT(thisDuck && thisDuck->name);

    if( thisDuck->vtable )
    {
        ASSERT( thisDuck->vtable->show );
        thisDuck->vtable->show(thisDuck);
    }
    else printf("\tHi! My name is %s.\n", thisDuck->name);
}

void
duckDestroy( p_Duck_t thisDuck )
{
    ASSERT(thisDuck && thisDuck->name);

    printf("\tDestroying Duck object with name: %s\n", thisDuck->name);
    if( thisDuck->vtable )
    {
        ASSERT( thisDuck->vtable->deinit && thisDuck->vtable->destroy );
        thisDuck->vtable->deinit(thisDuck);

        // Duck deinitialization
        memset(thisDuck->name, 0, MAX_CHARS_NAME);

        thisDuck->vtable->destroy(thisDuck);
    }
    else
    {
        memset(thisDuck->name, 0, MAX_CHARS_NAME);
        free(thisDuck);
    }
}